io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -618,10 +698,149 @@ schedule thread@Thread{
+      case maskst' of
+        -- If we're now unmasked then check for any pending async exceptions
+        Unmasked -> deschedule Interruptable thread' simstate
+        _        -> schedule                 thread' simstate
+
+    ThrowTo e tid' _ | tid' == tid -> do
+      -- Throw to ourself is equivalent to a synchronous throw,
+      -- and works irrespective of masking state since it does not block.
+      let thread' = thread { threadControl = ThreadControl (Throw e) ctl
+                           , threadMasking = MaskedInterruptible }
+      trace <- schedule thread' simstate
+      return (Trace time tid (EventThrowTo e tid) trace)
+
+    ThrowTo e tid' k -> do
+      let thread'   = thread { threadControl = ThreadControl k ctl }
+          willBlock = case Map.lookup tid' threads of
+                        Just t -> not (threadInterruptible t)
+                        _      -> False
+      if willBlock
+        then do
+          -- The target thread has async exceptions masked so we add the
+          -- exception and the source thread id to the pending async exceptions.
+          let adjustTarget t = t { threadThrowTo = (e, tid) : threadThrowTo t }
+              threads'       = Map.adjust adjustTarget tid' threads
+          trace <- deschedule Blocked thread' simstate { threads = threads' }
+          return $ Trace time tid (EventThrowTo e tid')
+                 $ Trace time tid EventThrowToBlocked
+                 $ trace
+        else do
+          -- The target thread has async exceptions unmasked, or is masked but
+          -- is blocked (and all blocking operations are interruptible) then we
+          -- raise the exception in that thread immediately. This will either
+          -- cause it to terminate or enter an exception handler.
+          -- In the meantime the thread masks new async exceptions. This will
+          -- be resolved if the thread terminates or if it leaves the exception
+          -- handler (when restoring the masking state would trigger the any
+          -- new pending async exception).
+          let adjustTarget t@Thread{ threadControl = ThreadControl _ ctl' } =
+                t { threadControl = ThreadControl (Throw e) ctl'
+                  , threadBlocked = False
+                  , threadMasking = MaskedInterruptible }
+              simstate'@SimState { threads = threads' }
+                         = snd (unblockThreads [tid'] simstate)
+              threads''  = Map.adjust adjustTarget tid' threads'
+              simstate'' = simstate' { threads = threads'' }
+
+          trace <- schedule thread' simstate''
+          return $ Trace time tid (EventThrowTo e tid')
+                 $ trace
+
+threadInterruptible :: Thread s a -> Bool
+threadInterruptible thread =
+    case threadMasking thread of
+      Unmasked                 -> True
+      MaskedInterruptible
+        | threadBlocked thread -> True  -- blocking operations are interruptible
```

Comment:
Be sure to conform to this http://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Exception.html#g:13

This part in particular

> The following operations are guaranteed not to be interruptible:
> ...
> takeMVar if the MVar is definitely full, and conversely putMVar if the MVar is definitely empty

Reply:
Yes, we comply with this. In particular the potentially blocking operations are STM txs (with retry) and `throwTo` (if the target is itself blocked). In both cases if the target thread is *actually* blocked (ie did `retry` or `throwTo` and blocked) so that the thread is not on the run queue, then it is blocked in an interruptable operation. And so in the `MaskedInterruptible` state we can interrupt it.

So perhaps to be a bit more precise than "blocking operations are interruptible" we mean that threads that are actually blocked are interruptable (because all operations that cause them to be blocked have this semantics). And that is the condition we test.
```
MaskedInterruptible
  | threadBlocked thread -> True
```